### ISSUE TITLE
Setting a /32 or /128 IP blockSize on a pool not possible

### ISSUE THREAD

**Issue by CallMeFoxie** [cilium/cilium#28637]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Similar to https://github.com/cilium/cilium/issues/24351 but in multi-pool mode we sometimes use /32 (or /128 for that matter) single-IP pools for stuff like reserved IPs or IPs with special use. However cilium does not use the first and last IP of the pool, meaning the smallest is /30 and still wastes at least two IPs per block. We use eBGP so the "first and last IPs" are of no concern to us.

Since the patch for LB-IPAM is in the works to improve this, the code should be reusable for normal pod IPAM as well.

### Cilium Version

1.15.0 branch

### Kernel Version

5.15.0

### Kubernetes Version

1.27.1

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by samip5** [comment]
Duplicate of #24351

**Comment by gandro** [comment]
I think it's very clear that this is something we want, as it has been requested quite often. It's less clear how to implement it. To summarize the discussion from Slack, there are a few concerns about using the current LB IPAM allocator as-is:

1. The new LB IPAM allocator works very fast when allocating "any" IP, but its way slower when allocating specific IPs, the more fragmented the IP space the longer it takes. Allocating specific IPs is performed during endpoint restoration at startup. However,  @dylandreimerink believes even in with heavy fragmentation allocating a specific IP is sub millisecond
2. The new LB IPAM allocator allocates IPs sequentially, and tries to re-use IPs if that reduces fragemenation. Where as the current multi-pool allocator allocates pod IPs at random. This means that the new allocator makes IP allocation much more predictable, allowing potential attackers to predict or reuse a specific pod IP. Random IP allocation does not fully solve that problem (after all, the IP space is typically small with ~250 IPs at most), but is probably still the preferred policies for pod IP allocation

**Comment by dylandreimerink** [comment]
To clarify for number 1, I just ran the benchmarks in the IP allocator. At 50000 blocks, it takes on average 0.3 ms to allocate a random specific IP, so not terrible but way slower than allocating "any" IP from the range. 

```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/ipalloc
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
BenchmarkHashAlloc_AllocPerStage/pre-alloced-50000-16         	    3985	    312188 ns/op	     166 B/op	       2 allocs/op
``` 

Allocating any IP takes roughly 200 nano seconds
```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/ipalloc
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
BenchmarkHashAlloc_AllocAny-16    	 6240414	       196.7 ns/op	       3 B/op	       0 allocs/op
PASS
ok  	github.com/cilium/cilium/pkg/ipalloc	1.495s
```

The reason behind this is that with any IP we just pick the first free block we can find and get an IP, so O(1). But for a specific IP we have to iterate over the list of free blocks to find the block containing the given IP, which is an O(n) op where n is the number of fragments/blocks of free IPs.

Lets say 0.3ms is the worst case, then we can still restore about 3000 IPs per second which would be a lot of pods for 1 node. So we might be good here. Even so, I think we can improve on this by batching a bunch of IPs if need be.

With regards to number 2. I suggest making a `Allocator[T].AllocRandom` for the use case of smaller ranges such as a /24. Max amount of fragments in that case is 128 which is likely not a huge deal.

For larger ranges such as IPv6 /48 (assuming we want to support that) we can make a third method that does a middle ground. It takes a target amount of fragments, lets say a 1000. If below the target we pick a random IP from the range which will likely increase the fragment count by 1. If at or above the fragment count we pick a random block, so 1 in the 1000, then take its start or end IP, so we have a randomness of 1 in 2000. Allocating from the start or end of a fragment doesn't add new fragments and sometimes removes one. We can change the target to get a different performance/security tradeoff. (not sure what to call this, semi-random?)

**Comment by CallMeFoxie** [comment]
out of curiosity, 200ns seems darn fast considering the allocation code is not called every second but only when a new pod spawns, no? I think that k8s scheduler itself is way slower than the alloc code.

And blocks are not /48 per ip pool I presume :) we use something like /104 at most I think? What about others?

**Comment by dylandreimerink** [comment]
> out of curiosity, 200ns seems darn fast considering the allocation code is not called every second but only when a new pod spawns, no?

Yes, its very fast. That is because its basically 2 pointer lookups, 1 number increment and 1 write to a hash map. The random number allocation will be slower.

> And blocks are not /48 per ip pool I presume :) we use something like /104 at most I think? What about others?

/104 is also 256 addresses so that would be great. When I hear people talk about IPv6 its always stupidly large ranges, since /48 is the smallest size you are allowed to announce on the internet. /104 is 16777216 IPs, so here having the semi-random allocation would be advisable to keep the fragments at a few thousand at most

**Comment by maxheyer** [comment]
Is there any news about this? As we have the same problems.

**Comment by samip5** [comment]
> Is there any news about this? As we have the same problems.

Not supported yet.

**Comment by fsbaraglia** [comment]
Hi all,

I am using this configuration for the moment.
I created an ippool with this configuration and selector based on label:
apiVersion: "cilium.io/v2alpha1"
kind: CiliumLoadBalancerIPPool
metadata:
  name: "blue-pool"
spec:
  cidrs:
  - cidr: "172.16.2.90/30"
  serviceSelector:
    matchLabels:
      color: blue
      
      
and I am adding the annotation in the svc
annotations:
    io.cilium/lb-ipam-ips: 172.16.2.90
labels:
    color: blue

In this case, I am controlling with  (  serviceSelector: matchLabels: colour: blue ) to which service the IP will be assigned and which IP will be assigned.
I know that this is a bit annoying if the loadbalancer's IP changes, but maybe you can trick the configuration into adapting if the IP changes via IaC.

**Comment by CallMeFoxie** [comment]
We have "developed" a small patch that we keep in our repo internally. Basically it removes the "do not use first and last IP from pool" logic, it removes "ip pool is 2^n-2" logic and several other things (I think it was something like 4 places it needed to be patched out) so we are fine right now and happily using it. Hacky but it works and allows us to actually use cilium in production clusters and do things we would not be able to do without it.

@fsbaraglia Our problem is not with ingress to the pod but with egress from that pod :) we have tied special ACLs on routers and firewalls to specific ranges and do not want 99.99% of our pods use these ranges by default.

(And no, we cannot connect those routers and firewalls into clustermesh :D)

**Comment by fsbaraglia** [comment]
Ah ok, good to know @CallMeFoxie 
I guess I missed the egress context in the previous messages.

**Comment by nayutah** [comment]
> We have "developed" a small patch that we keep in our repo internally. Basically it removes the "do not use first and last IP from pool" logic, it removes "ip pool is 2^n-2" logic and several other things (I think it was something like 4 places it needed to be patched out) so we are fine right now and happily using it. Hacky but it works and allows us to actually use cilium in production clusters and do things we would not be able to do without it.
> 
@CallMeFoxie We are working on a unified operator for multiple databases, the project is called KubeBlocks, we also encountered the same problem as some databases assuming that each IP address has a network device, they rely on the network device (like UUID) as their identity, it is useful in data replication(like paxos) for telling others in the replication group whether it is a new candidate or an old one just recovered from a failure through their identity, in such case, the static 32/128 pod IP seems the  possible and easy way to solve the problem without changing the codes of replication. I am interested that if your fixed-ip cilium branch is open-source or will you push it back to the upstream, or anything else...   Thank you for your attention and glad to see your progress.

**Comment by CRASH-Tech** [comment]
not stale

**Comment by oblazek** [comment]
@nayutah sorry, we have not made it yet into upstream with @CallMeFoxie.. we still keep our "modification" locally. @gandro sorry we've not been active much for the past 2 months... have you discussed anything internally regarding this issue?

**Comment by oblazek** [comment]
just to sum it up.. we are talking about multipool IPAM, but being able to specify /32 or /128 address as `maskSize` in `CiliumPodIPPool`

**Comment by gandro** [comment]
> @gandro sorry we've not been active much for the past 2 months... have you discussed anything internally regarding this issue?

I don't think there has been any update on this recently. I believe the above comments still hold true. I'm not aware of anyone actively working on this.

**Comment by CallMeFoxie** [comment]
noooo not stale!

**Comment by juliusmh** [comment]
@CallMeFoxie mind sharing the patches required for Cilium to work in that way?

**Comment by CallMeFoxie** [comment]
> @CallMeFoxie mind sharing the patches required for Cilium to work in that way?

cc @oblazek can you pull the required commits from the git? :)

**Comment by oblazek** [comment]
this is what we've got:
```
diff --git a/pkg/ipam/allocator/multipool/pool_allocator.go b/pkg/ipam/allocator/multipool/pool_allocator.go
index f0ae28eff5..d22fe75b92 100644
--- a/pkg/ipam/allocator/multipool/pool_allocator.go
+++ b/pkg/ipam/allocator/multipool/pool_allocator.go
@@ -71,9 +71,6 @@ func addrsInPrefix(p netip.Prefix) *big.Int {
        // compute number of addresses in prefix, i.e. 2^bits
        addrs := new(big.Int)
        addrs.Lsh(big.NewInt(1), uint(p.Addr().BitLen()-p.Bits()))
-       // subtract network and broadcast address, which are not available for
-       // allocation in the cilium/ipam library for now
-       addrs.Sub(addrs, big.NewInt(2))
        if addrs.Sign() < 0 {
                return big.NewInt(0)
        }
diff --git a/pkg/ipam/service/ipallocator/allocator.go b/pkg/ipam/service/ipallocator/allocator.go
index bd79d352b7..da9c4b7336 100644
--- a/pkg/ipam/service/ipallocator/allocator.go
+++ b/pkg/ipam/service/ipallocator/allocator.go
@@ -68,11 +68,11 @@ type Range struct {
 // the backing store.
 func NewCIDRRange(cidr *net.IPNet) *Range {
        base := bigForIP(cidr.IP)
-       max := maximum(0, int(RangeSize(cidr)-2)) // don't use the network broadcast,
+       max := maximum(0, int(RangeSize(cidr)))
 
        return &Range{
                net:   cidr,
-               base:  base.Add(base, big.NewInt(1)), // don't use the network base
+               base:  base.Add(base, big.NewInt(0)),
                max:   max,
                alloc: allocator.NewAllocationMap(int(max), cidr.String()),
        }
```

**Comment by bewing** [comment]
Is there an actual deployment method where the first and last IPs are not usable?  I would expect a more robust method of marking IPs as unusable if that were the case, as one might have to mark an IP as reserved for a router or similar.

**Comment by gandro** [comment]
FYI: PR by @juliusmh will remove the use of "first and last" IPs for /32 or /128 CIDRs: https://github.com/cilium/cilium/pull/34618 - 

It will continue to reserve the first and last IP for larger CIDRs (i.e. there is no `allowFirstLastIPs` option like in `CiliumLoadBalancerIPPool`), but since /32 and /128 are unusable otherwise, this seems like a reasonable stop-gap solution even without changing the CRD.

**Comment by chaeynz** [comment]
@CallMeFoxie i think this can be closed

**Comment by gandro** [comment]
I agree that the core issue has been addressed. We also now internally also have `WithAllowFirstLastIPs` in the allocator thanks to Hadrien's work. 

I don't think we are exposing `WithAllowFirstLastIPs` in the CRD yet, however. It would be nice to have that too. I'll create a new issue for that: https://github.com/cilium/cilium/issues/46201
